iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -0,0 +1,218 @@
+
+static ParseResult parseCommonStructuredOpParts(
+    OpAsmParser &parser, OperationState &result,
+    SmallVectorImpl<Type> &inputTypes, SmallVectorImpl<Type> &outputTypes) {
+  llvm::SMLoc inputsOperandsLoc, outputsOperandsLoc;
+  SmallVector<OpAsmParser::OperandType, 4> inputsOperands, outputsOperands;
+
+  parser.parseOptionalAttrDict(result.attributes);
+
+  if (succeeded(parser.parseOptionalKeyword("ins"))) {
+    if (parser.parseLParen()) return failure();
+
+    inputsOperandsLoc = parser.getCurrentLocation();
+    if (parser.parseOperandList(inputsOperands) ||
+        parser.parseColonTypeList(inputTypes) || parser.parseRParen())
+      return failure();
+  }
+
+  if (succeeded(parser.parseOptionalKeyword("outs"))) {
+    outputsOperandsLoc = parser.getCurrentLocation();
+    if (parser.parseLParen() || parser.parseOperandList(outputsOperands) ||
+        parser.parseColonTypeList(outputTypes) || parser.parseRParen())
+      return failure();
+  }
+
+  if (parser.resolveOperands(inputsOperands, inputTypes, inputsOperandsLoc,
+                             result.operands) ||
+      parser.resolveOperands(outputsOperands, outputTypes, outputsOperandsLoc,
+                             result.operands))
+    return failure();
+
+  result.addAttribute("operand_segment_sizes",
+                      parser.getBuilder().getI32VectorAttr(
+                          {static_cast<int32_t>(inputsOperands.size()),
+                           static_cast<int32_t>(outputsOperands.size())}));
+  return success();
+}
+
+static ParseResult parseNamedStructuredOpResults(
+    OpAsmParser &parser, SmallVectorImpl<Type> &resultTypes) {
+  if (parser.parseOptionalArrowTypeList(resultTypes)) return failure();
+  return success();
+}
+
+template <typename NamedStructuredOpType>
+static void printCommonStructuredOpParts(OpAsmPrinter &p,
+                                         NamedStructuredOpType op) {
+  if (!op.inputs().empty()) {
+    p << " ins(" << op.inputs() << " : " << op.inputs().getTypes() << ")";
+  }
+  if (!op.outputs().empty()) {
+    p << " outs(" << op.outputs() << " : " << op.outputs().getTypes() << ")";
+  }
+}
+
+static void printNamedStructuredOpResults(OpAsmPrinter &p,
+                                          TypeRange resultTypes) {
+  if (resultTypes.empty()) return;
+  p.printOptionalArrowTypeList(resultTypes);
+}
+
+//===----------------------------------------------------------------------===//
+// SortOp
+//===----------------------------------------------------------------------===//
+
+void SortOp::getEffects(
+    SmallVectorImpl<SideEffects::EffectInstance<MemoryEffects::Effect>>
+        &effects) {
+  SmallVector<Value> inputBuffers = getInputBufferOperands();
+  SmallVector<Value> outputBuffers = getOutputBufferOperands();
+  getEffectsImpl(effects, getOperation()->getResults(), inputBuffers,
+                 outputBuffers);
+}
+
+static ParseResult parseSortOp(OpAsmParser &parser, OperationState &result) {
```

Comment:
Oh, is there no way to use the [declarative assembly format](https://mlir.llvm.org/docs/OpDefinitions/#declarative-assembly-format). Makes it much easier to understand whats happening.
Not needed for this PR per-se but an change to Linalg upstream to add a custom directive would make this much nicer.

Reply:
Good point! I will revisit if we can use the declarative assembly format. This is basically following Linalg implementation.